README.md
[![Code Climate](https://lzw.me/images/logo.png)](https://lzw.me)
[![@lzwme/node-exchange-mailer](https://nodei.co/npm/@lzwme/node-exchange-mailer.png)](npm-url)

# node-exchange-mailer

[![NPM version][npm-badge]][npm-url]
[![node version][node-badge]][node-url]
[![npm download][download-badge]][download-url]
[![GitHub issues][issues-badge]][issues-url]
[![GitHub forks][forks-badge]][forks-url]
[![GitHub stars][stars-badge]][stars-url]
[![minzipped size][bundlephobia-badge]][bundlephobia-url]

使用 nodejs 发送基于 EWS(Microsoft Exchange Web Serveice) 邮件服务的邮件。

基于 `node-ews` 封装的 EWS 邮件发送 API 方法。

## 安装

```bash
# use npm
npm add @lzwme/node-exchange-mailer
# or use yarn
yarn add @lzwme/node-exchange-mailer
```

## 使用示例

### 1. 示例一：使用账号和密码发送邮件

```js
const { sendMailByEws } = require('@lzwme/node-exchange-mailer');

const options = {
  subject: '[ews]邮件主题测试',
  html: 'HTML 格式内容，<b>优先级高于 text</b>',
  /** TEXT 文本格式邮件正文内容(优先级低于 html 参数) */
  text: 'text',
  /** 收件人列表 */
  to: 'xa@lzw.me',
  /** 抄送人列表 */
  cc: 'xb@lzw.me, xc@lzw.me',
  /** 密送人列表 */
  bcc: 'xd@lzw.me, xe@lzw.me',
  // text: '纯文本格式内容',
  ewsConfig: {
    username: 'myuser@lzw.me',
    password: 'mypassword',
    host: 'https://ews.lzw.me',
  },
};
sendMailByEws(options).then((result) => console.log(result));
```

### 2. 示例二：使用 `ntlm` 授权模式发送邮件，避免明文密码被直接泄漏

根据明文密码生成 `ntlm` 认证模式所需的密钥：

```js
const { genNtlmHashedPwd } = require('@lzwme/node-exchange-mailer');

const password = 'mypassword'; // process.argv.slice(2)[0];
const { nt_password, lm_password } = genNtlmHashedPwd(password, true);
// => password: mypassword
// => nt_password: a991ae45aa987a1a48c8bdc1209ff0e7
// => lm_password: 74ac99ca40ded420dc1a73e6cea67ec5
```

配置 `nt_password` 和 `lm_password` 字段，基于 ntlm 认证模式发送邮件：

```js
const { sendMailByEws } = require('@lzwme/node-exchange-mailer');

const options = {
  subject: '[ews]邮件 ntlm 测试',
  to: 'xa@lzw.me',
  html: 'HTML 格式内容，<b>优先级高于 text</b>',
  ewsConfig: {
    host: 'https://ews.lzw.me',
    username: 'myuser@lzw.me',
    /** 密码加密后的秘钥(NTLMAuth.nt_password)。为字符串时，应为 hex 编码结果 */
    nt_password: 'a991ae45aa987a1a48c8bdc1209ff0e7',
    /** 密码加密后的秘钥(NTLMAuth.lm_password)。为字符串时，应为 hex 编码结果 */
    lm_password: '74ac99ca40ded420dc1a73e6cea67ec5',
  },
};

sendMailByEws(options).then((result) => console.log(result));
```

### 示例三：发送带附件的邮件

```js
const { sendMailByEws } = require('@lzwme/node-exchange-mailer');
const path = require('path');

const options = {
  subject: '[ews]邮件附件测试',
  html: [
    `HTML 格式内容，<b>包含附件</b>: <a href="https://lzw.me">https://lzw.me</a>`,
    `<img src='cid:png01' style='width:144px;height:auto'>`,
  ].join('<br>'),
  to: 'xa@lzw.me',
  ewsConfig: {
    host: 'https://ews.lzw.me',
    username: 'myuser@lzw.me',
    nt_password: 'a991ae45aa987a1a48c8bdc1209ff0e7',
    lm_password: '74ac99ca40ded420dc1a73e6cea67ec5',
  },
  soapHeader: {
    't:RequestServerVersion': {
      attributes: {
        Version: 'Exchange2013_SP1', // Exchange2007<_SP1>, Exchange2010<_SP1>, , Exchange2013<_SP1>
      },
    },
  },
  /**
   * 附件列表
   * {@link https://github.com/CumberlandGroup/node-ews/issues/40 | 发送附件参考 }
   */
  attachments: [
    {
      /** 附件名称 */
      Name: 'lzw.me.txt',
      /** 附件类型(如文本： text/plain) */
      ContentType: 'text/plain',
      /** 附件内容(请使用 base64 编码) */
      Content: 'ZmlsZSBhdHRhY2htZW50LiAtIHRlc3QgYnkgaHR0cHM6Ly9sencubWU=',
      IsInline: false,
    },
    {
      /** 附件名称 */
      Name: 'lzwme.png',
      /** 附件类型(如文本： text/plain) */
      ContentType: 'image/png',
      /** 附件内容(使用 base64 编码) */
      Content: readFileSync(path.resolve(__dirname, './lzwme-144x144.png'), {
        encoding: 'base64',
      }),
      IsContactPhoto: false,
      IsInline: true,
      ContentId: 'png01',
    },
  ],
};

sendMailByEws(options).then((result) => console.log(result));
```

### 示例四：发送 smtp 协议的邮件

本仓库基于 `nodemailer` 简单封装了一个 smtp 协议发送邮件的方法。用法如下：

安装 `nodemailer` 依赖：

```bash
yarn add nodemailer
```

发送 smtp 协议的邮件示例：

```js
const { sendMailBySmtp } = require('@lzwme/node-exchange-mailer');
const path = require('path');

const options = {
  subject: '[smtp]Subject for test',
  /** HTML 格式邮件正文内容 */
  html: [
    `email content for test: <a href="https://lzw.me">https://lzw.me</a>`,
    `<img src='cid:png01' style='width:144px;height:auto'>`,
  ].join('<br>'),
  /** TEXT 文本格式邮件正文内容 */
  text: '',
  to: 'l@lzw.me',
  /** 附件列表 */
  attachments: [
    {
      filename: 'lzwme.png',
      cid: 'png01',
      path: path.resolve(__dirname, 'lzwme-144x144.png'),
    },
    {
      filename: 'lzwme.txt',
      content: '邮件内容，test by <b>lzw.me</b>',
    },
  ],
};
const smtpConfig = {
  // service: 'qq',
  host: 'smtp.qq.com', // QQ: smtp.qq.com; 网易: smtp.163.com
  port: 465, // 端口号。QQ邮箱  465，网易邮箱 25
  secure: true,
  auth: {
    user: 'xxx@qq.com', // 邮箱账号
    pass: 'xxxxxx', // 邮箱的授权码
  },
};
sendMailBySmtp(options, smtpConfig).then((result) => console.log(result));
```

## 扩展参考

- [使用 node.js 发送基于 STMP 与 MS Exchange 邮件的方法](https://lzw.me/a/nodejs-stmp-exchange-email-send.html)
- [EWS reference for Exchange](https://docs.microsoft.com/en-us/exchange/client-developer/web-service-reference/ews-reference-for-exchange)
- [createitem-operation-email-message](https://docs.microsoft.com/en-us/exchange/client-developer/web-service-reference/createitem-operation-email-message)

## License

`@lzwme/node-exchange-mailer` is released under the MIT license.

该插件由[志文工作室](https://lzw.me)开发和维护。

[stars-badge]: https://img.shields.io/github/stars/lzwme/node-exchange-mailer.svg
[stars-url]: https://github.com/lzwme/node-exchange-mailer/stargazers
[forks-badge]: https://img.shields.io/github/forks/lzwme/node-exchange-mailer.svg
[forks-url]: https://github.com/lzwme/node-exchange-mailer/network
[issues-badge]: https://img.shields.io/github/issues/lzwme/node-exchange-mailer.svg
[issues-url]: https://github.com/lzwme/node-exchange-mailer/issues
[npm-badge]: https://img.shields.io/npm/v/@lzwme/node-exchange-mailer.svg?style=flat-square
[npm-url]: https://npmjs.org/package/@lzwme/node-exchange-mailer
[node-badge]: https://img.shields.io/badge/node.js-%3E=_10.0.0-green.svg?style=flat-square
[node-url]: https://nodejs.org/download/
[download-badge]: https://img.shields.io/npm/dm/@lzwme/node-exchange-mailer.svg?style=flat-square
[download-url]: https://npmjs.org/package/@lzwme/node-exchange-mailer
[bundlephobia-url]: https://bundlephobia.com/result?p=@lzwme/node-exchange-mailer@latest
[bundlephobia-badge]: https://badgen.net/bundlephobia/minzip/@lzwme/node-exchange-mailer@latest


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

## 1.1.0 (2021-01-23)


### Features

* 增加基于 nodemialer 的 smtp 邮件发送方法；补全说明文档 ([769bea5](https://github.com/lzwme/node-exchange-mailer/commit/769bea5aa00bc05a16df7fc24d9d7c130e0726c9))
* 支持发送附件、发件人标识代理等更多参数配置 ([fd53ff2](https://github.com/lzwme/node-exchange-mailer/commit/fd53ff2b36c5b0ee4266f0296a5054a0862ba3f4))


test/mail-smtp.ts
process.env.DEBUG = '*';

import { sendMailBySmtp } from '../src';

const opts = {
  subject: 'subject for test',
  /** HTML 格式邮件正文内容 */
  html: `email content for test: <a href="https://lzw.me">https://lzw.me</a>`,
  /** TEXT 文本格式邮件正文内容 */
  text: '',
  to: 'xxx@lzw.me',
  // 附件列表
  // attachments: [],
};
const smtpConfig = {
  host: 'smtp.qq.com', // QQ: smtp.qq.com; 网易: smtp.163.com
  port: 465, // 端口号。QQ邮箱  465，网易邮箱 25
  secure: true,
  auth: {
    user: 'xxx@qq.com', // 邮箱账号
    pass: 'xxxxxx', // 邮箱的授权码
  },
};
sendMailBySmtp(opts, smtpConfig).then((result) => console.log(result));


test/smtp-test.ts
process.env.DEBUG = '*';

import path from 'path';
import { sendMailBySmtp, ISmtpMsgOptions, ISmtpConfig } from '../src';
import testConfig from '../test.config';

const options: ISmtpMsgOptions = {
  subject: '[smtp]Subject for test',
  /** HTML 格式邮件正文内容 */
  html: [
    `email content for test: <a href="https://lzw.me">https://lzw.me</a>`,
    `<img src='cid:png01' style='width:144px;height:auto'>`,
  ].join('<br>'),
  /** TEXT 文本格式邮件正文内容 */
  text: '',
  to: 'xxx@lzw.me',
  /** 附件列表 */
  attachments: [
    {
      filename: 'lzwme.png',
      cid: 'png01',
      path: path.resolve(__dirname, 'lzwme-144x144.png'),
    },
    {
      filename: 'lzwme.txt',
      content: '邮件内容，test by <b>lzw.me</b>',
    },
  ],
  ...testConfig.smtpOptions,
};
const smtpConfig: ISmtpConfig = {
  service: 'qq',
  // host: 'smtp.qq.com', // QQ: smtp.qq.com; 网易: smtp.163.com
  // port: 465, // 端口号。QQ邮箱  465，网易邮箱 25
  // secure: true,
  // auth: {
  //   user: 'xxx@qq.com', // 邮箱账号
  //   pass: 'xxxxxx', // 邮箱的授权码
  // },
  ...testConfig.smtpConfig,
};

sendMailBySmtp(options, smtpConfig).then((result) => console.log(result));


test/get-ntlm-pwd.ts
process.env.DEBUG = '*';

import { genNtlmHashedPwd } from '../src';

const password = process.argv.slice(2)[0] || 'mypassword';
genNtlmHashedPwd(password, true);


src/lib/utils.ts
/*
 * @Author: lzw
 * @Date: 2021-01-14 22:10:23
 * @LastEditors: lzw
 * @LastEditTime: 2021-01-15 11:01:08
 * @Description: 工具方法
 */

import debug from 'debug';

export interface PlainObject {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  [key: string]: any;
}

/** debug log */
export const log = debug('ews:nem');

export function toMailAddressList(to: string | string[]): string[] {
  const toArr = [];
  if (typeof to === 'string') to = to.split(',').map((d) => d.trim());
  if (Array.isArray(to)) toArr.push(...to);
  return toArr.filter((d) => typeof d === 'string' && /.+@.+/.test(d));
}

/** 移除给定对象所有子级属性值为空的节点 */
export function delEmptyKeys(obj: PlainObject | unknown[]) {
  if (!obj) return obj;
  if (Array.isArray(obj)) {
    obj.forEach((d) => delEmptyKeys(d));
    return obj;
  }

  if (typeof obj === 'object') {
    Object.keys(obj).forEach((key) => {
      const item = obj[key];

      if (item == null || item === '') delete obj[key];
      else if (Array.isArray(item)) {
        if (!item.length) delete obj[key];
        else delEmptyKeys(item);
      } else if (typeof item === 'object') {
        delEmptyKeys(item);
        if (!Object.keys(item).length) delete obj[key];
      }
    });
  }

  return obj;
}


src/lib/send-mail-smtp.ts
import nodemailer from 'nodemailer';
import { PlainObject, toMailAddressList } from './utils';

/**
 * sendMail(data) 发送邮件的数据配置
 * {@link https://nodemailer.com/message/ | 详细参考}
 */
export interface ISmtpMsgOptions extends PlainObject {
  from?: string;
  replayTo?: string;
  sender?: string;
  /** 邮件主题 */
  subject?: string;
  /** HTML 格式邮件正文内容 */
  html?: string;
  /** TEXT 文本格式邮件正文内容(优先级低于 html 参数) */
  text?: string;
  /** 收件人列表 */
  to?: string | string[];
  /** 抄送人列表 */
  cc?: string | string[];
  /** 暗抄送人列表 */
  bcc?: string | string[];
  envelope?: {
    from?: string;
    to?: string;
  };
  /** 附件内容 */
  attachments?: {
    filename?: string;
    path?: string;
    content?: string | Buffer;
    /** 在邮件正文中的引用 ID */
    cid?: string;
  }[];
}

export interface ISmtpConfig extends PlainObject {
  auth: {
    /** login or oauth2 */
    type?: string;
    /** 邮箱账号 */
    user?: string;
    /** 邮箱的授权码 */
    pass?: string;
    accessToken?: string;
  } & PlainObject;
  /** smtp 服务地址。如 QQ: smtp.qq.com; 网易163: smtp.163.com */
  host?: string;
  /** 端口号。QQ邮箱  465，网易邮箱 25 */
  port?: number;
  secure?: boolean;
  pool?: boolean;
  tls?: PlainObject;
  sevice?: string;
}

/**
 * 基于 smtp 协议发送邮件
 */
export async function sendMailBySmtp(options: ISmtpMsgOptions, smtpConfig: ISmtpConfig) {
  const resultInfo = {
    /** 状态码。 为 0 表示成功，否则表示失败 */
    code: 0,
    /** 结果提示消息 */
    msg: '',
    /** 邮件发送返回的原始结果 */
    result: null as PlainObject,
  };
  const mailOpts = Object.assign(
    {
      from: `Notify <${smtpConfig.auth.user}>`,
      subject: 'Notify',
      // attachments: [{
      //     filename: 'data1.json',
      //     path: path.resolve(__dirname, 'data1.json')
      // }, {
      //     filename: 'pic01.jpg',
      //     path: path.resolve(__dirname, 'pic01.jpg')
      // }, {
      //     filename: 'test.txt',
      //     path: path.resolve(__dirname, 'test.txt')
      // }],
    } as ISmtpMsgOptions,
    options
  );

  mailOpts.to = toMailAddressList(mailOpts.to);

  if (!smtpConfig.auth) {
    resultInfo.code = -1000;
    resultInfo.msg = '未配置邮件发送信息';
  } else if (!mailOpts.to.length) {
    resultInfo.code = -1002;
    resultInfo.msg = 'Options.to can not be null';
  }

  if (resultInfo.code) return resultInfo;

  const transporter = nodemailer.createTransport(smtpConfig);
  try {
    const info = await transporter.sendMail(mailOpts);

    resultInfo.code = info.responseCode || info.code || 0;
    resultInfo.msg = info.response;
    resultInfo.result = info;
  } catch (error) {
    console.log(error);
    resultInfo.code = error.responseCode || error.code || 9001;
    resultInfo.msg = error.response || error.message || error;
  }

  return resultInfo;
}


src/lib/send-mail.ts
/*
 * @Author: lzw
 * @Date: 2021-01-14 21:35:47
 * @LastEditors: lzw
 * @LastEditTime: 2021-01-23 18:50:12
 * @Description: send exchange mail by node-ews
 */
import EWS from 'node-ews';
import { delEmptyKeys, log, PlainObject, toMailAddressList } from './utils';

interface ISoapHeader extends PlainObject {
  't:RequestServerVersion'?: { attributes: { Version: string } };
}

/**
 * 附件项格式
 *
 * {@link https://docs.microsoft.com/en-us/exchange/client-developer/web-service-reference/attachments-ex15websvcsotherref | 相关文档参考 }
 * {@link https://github.com/CumberlandGroup/node-ews/issues/40 | 发送附件参考 }
 */
export interface IAttachmentItem extends PlainObject {
  /** 附件名称 */
  Name: string;
  /** 附件扩展（MIME）类型(如文本： text/plain，png图片： image/png) */
  ContentType: string;
  /** 附件内容(请使用 base64 编码) */
  Content: string; // 'VGhpcyBpcyB0aGUgc2Vjb25kIGZpbGUgYXR0YWNobWVudC4=';
  /** 附件唯一标记，可在邮件正文中引用 */
  ContentId?: string;
  /** 包含与附件内容的位置相对应的统一资源标识符（URI） */
  ContentLocation?: string;
  /** 是否为联系人图片 */
  IsContactPhoto?: boolean;
  /** 是否内联显示 */
  IsInline?: boolean;
}
export interface IEwsSendOptions {
  /** 邮件主题 */
  subject?: string;
  /** HTML 格式邮件正文内容 */
  html?: string;
  /** TEXT 文本格式邮件正文内容(优先级低于 html 参数) */
  text?: string;
  /** 收件人列表 */
  to?: string | string[];
  /** 抄送人列表 */
  cc?: string | string[];
  /** 密送人列表 */
  bcc?: string | string[];
  /** 回复人列表 */
  replyTo?: string | string[];
  from?: string;
  /** 标识项目的发件人 */
  sender?: string;
  receivedBy?: string;
  /** 标识代理访问方案中的主体。与代理访问方案中的 From 和 ReceivedBy 元素一起使用 */
  receivedRepresenting?: string;
  /** 附件列表
   *
   * {@link https://docs.microsoft.com/en-us/exchange/client-developer/web-service-reference/attachments-ex15websvcsotherref | 相关文档参考 }
   * {@link https://github.com/CumberlandGroup/node-ews/issues/40 | 发送附件参考 }
   */
  attachments?: IAttachmentItem[];
  /**
   * 其他 Message 子属性配置。用于扩展不同版本可能存在的差异字段。具体可通过 `https://<ews服务器host>/ews/types.xsd` 的 Message 类型定义查询
   *
   * {@link https://docs.microsoft.com/en-us/exchange/client-developer/web-service-reference/message-ex15websvcsotherref | 官方文档相关格式说明参考}
   */
  messageExtends?: PlainObject;
  /** ewsArgs.Item 其他属性扩展配置。用于配置扩展。例如增加发送附件相关的配置 */
  itemExtends?: PlainObject;
  /** node-ews: 配置初始化参数 */
  ewsConfig?: {
    /** Exchange 服务地址 */
    host?: string;
    /** 用户名 */
    username?: string;
    /** 密码 */
    password?: string;
    /** 密码加密后的秘钥(NTLMAuth.nt_password)。为字符串时，应为 hex 编码结果 */
    nt_password?: string | Buffer;
    /** 密码加密后的秘钥(NTLMAuth.lm_password)。为字符串时，应为 hex 编码结果 */
    lm_password?: string | Buffer;
    /** 基于 auth=bearer 授权模式时的 token */
    token?: string;
    /** 授权模式。默认为 ntlm */
    auth?: 'basic' | 'bearer' | 'ntlm';
    /** 指定临时文件目录 */
    temp?: string;
  } & PlainObject;
  /** node-ews: EWS 第二个参数，视不同 auth 类型而不同 */
  ewsOptions?: PlainObject;
  /** node-ews soapHeader */
  soapHeader?: ISoapHeader | ((method, location, soapAction, args) => ISoapHeader);
}

/**
 *  发送 Exchange(EWS) 邮件
 */
export async function sendMailByEws(options: IEwsSendOptions) {
  const resultInfo = {
    /** 状态码。 为 0 表示成功，否则表示失败 */
    code: 0,
    /** 结果提示消息 */
    msg: '',
    /** 邮件发送返回的原始结果 */
    result: null as PlainObject,
  };
  /** 收件人列表 */
  const toRecipients = toMailAddressList(options.to);
  const ewsConfig: IEwsSendOptions['ewsConfig'] = Object.assign(
    {
      // auth: 'ntlm',
    },
    options.ewsConfig
  );

  if (!ewsConfig.username || (!ewsConfig.password && !ewsConfig.lm_password)) {
    resultInfo.code = -1001;
    resultInfo.msg = 'Options.ewsConfig.uername or Options.ewsConfig.password can not be null';
  } else if (!toRecipients.length) {
    resultInfo.code = -1002;
    resultInfo.msg = 'Options.to can not be null';
  }

  if (resultInfo.code) return resultInfo;

  ['nt_password', 'lm_password'].forEach((key) => {
    if (ewsConfig[key] && typeof ewsConfig[key] === 'string') {
      ewsConfig[key] = Buffer.from(ewsConfig[key], 'hex');
    }
  });

  Object.keys(ewsConfig).forEach((key) => {
    if (!ewsConfig[key]) delete ewsConfig[key];
  });

  const ews = new EWS(ewsConfig, options.ewsOptions);
  /** ews api function */
  const ewsFunction = 'CreateItem';
  /** ews api function args */
  const ewsArgs = {
    attributes: {
      MessageDisposition: 'SendAndSaveCopy',
    },
    SavedItemFolderId: {
      DistinguishedFolderId: {
        attributes: {
          Id: 'sentitems',
        },
      },
    },
    Items: {
      Message: {
        ItemClass: 'IPM.Note',
        Subject: options.subject,
        Body: {
          attributes: {
            BodyType: options.html ? 'HTML' : 'Text',
          },
          $value: options.html || options.text,
        },
        From: {
          Mailbox: toMailAddressList(options.from).map((to) => ({ EmailAddress: to })),
        },
        Sender: {
          Mailbox: toMailAddressList(options.sender).map((to) => ({ EmailAddress: to })),
        },
        ReplyTo: {
          Mailbox: toMailAddressList(options.replyTo).map((to) => ({ EmailAddress: to })),
        },
        ReceivedBy: {
          Mailbox: toMailAddressList(options.receivedBy).map((to) => ({ EmailAddress: to })),
        },
        ReceivedRepresenting: toMailAddressList(options.receivedRepresenting).map((to) => ({ EmailAddress: to })),
        ToRecipients: {
          Mailbox: toMailAddressList(options.to).map((to) => ({ EmailAddress: to })),
        },
        CcRecipients: {
          Mailbox: toMailAddressList(options.cc).map((to) => ({ EmailAddress: to })),
        },
        BccRecipients: {
          Mailbox: toMailAddressList(options.bcc).map((to) => ({ EmailAddress: to })),
        },
        IsRead: 'false',
        Attachments: {
          FileAttachment: options.attachments,
        },
      },
    },
  };

  if (options.itemExtends) {
    if (options.itemExtends.Message) {
      console.warn('请通过 options.messageExtends 参数配置 Message 扩展信息', options.itemExtends.Message);
      delete options.itemExtends.Message;
    }

    Object.assign(ewsArgs.Items, options.itemExtends);
  }
  if (options.messageExtends) Object.assign(ewsArgs.Items.Message, options.messageExtends);

  // 移除部分空值属性
  delEmptyKeys(ewsArgs);
  log(JSON.stringify(ewsArgs, null, 2));

  try {
    const result = await ews.run(ewsFunction, ewsArgs, options.soapHeader);
    resultInfo.result = result;
    log('send mail result:\n', JSON.stringify(result));

    const cirMsg = result.ResponseMessages.CreateItemResponseMessage;
    if (cirMsg) {
      if (cirMsg.MessageText) resultInfo.msg = cirMsg.MessageText;
      if (cirMsg.ResponseCode !== 'NoError') resultInfo.code = cirMsg.ResponseCode;
    }
  } catch (err) {
    console.log(err.message || 'ERROR:', '\n\n', err.stack);
    resultInfo.code = err.code || 9001;
    resultInfo.msg = err.message || err.stack;
  }

  return resultInfo;
}
